crates/sui-rosetta/src/types.rs

```diff
@@ -947,6 +959,63 @@ impl InternalOperation {
                 builder.pay_sui(recipients, amounts)?;
                 builder.finish()
             }
+            Self::PayCoin {
+                recipients,
+                amounts,
+                ..
+            } => {
+                let mut builder = ProgrammableTransactionBuilder::new();
+
+                let mut coins = metadata.objects.clone().into_iter();
+                let Some(coin) = coins.next() else {
+                    return Err(Error::InvalidInput("coins vector is empty".to_string()));
+                };
+                let coin_arg = builder.obj(ObjectArg::ImmOrOwnedObject(coin))?;
+                let merge_args: Vec<_> = coins
+                    .map(|c| builder.obj(ObjectArg::ImmOrOwnedObject(c)))
+                    .collect::<Result<_, _>>()?;
+                if !merge_args.is_empty() {
+                    builder.command(Command::MergeCoins(coin_arg, merge_args));
+                }
+                if recipients.len() != amounts.len() {
+                    return Err(Error::InvalidInput(
+                "Recipients and amounts mismatch. Got {recipients.len()} recipients but {amounts.len()} amounts".to_string()));
+                }
+
+                // collect recipients in the case where they are non-unique in order
+                // to minimize the number of transfers that must be performed
+                let mut recipient_map: IndexMap<SuiAddress, Vec<usize>> = IndexMap::new();
+                let mut amt_args = Vec::with_capacity(recipients.len());
+                for (i, (recipient, amount)) in recipients.into_iter().zip(amounts).enumerate() {
+                    recipient_map.entry(recipient).or_default().push(i);
+                    amt_args.push(builder.pure(amount)?);
+                }
+                let Argument::Result(split_primary) =
+                    builder.command(Command::SplitCoins(coin_arg, amt_args))
+                else {
+                    panic!("self.command should always give a Argument::Result")
```

Comment:
I think you'd want an error, not a panic?

Reply:
yep that's jsut a copy from the pay method, the PR is draft will take care of these details once I make the flow work.

Reply:
The thing is that `pay` uses the native PTB transfers, while you're trying to port the transfer logic into a move call `0x2::transfer::public_transfer` (hence adding the requirement for type params too). 

The only difference between a SUI transfer would be that you cannot use "gas" to do splits, so I don't think that should be that way.